src/interp/interp.cc

```diff
@@ -1833,7 +1833,7 @@ RunResult Thread::StepInternal(Trap::Ptr* out_trap) {
       return DoSimdBinop(SaturatingRoundingQMul<s16>);
 
     case O::I32X4DotI16X8S: return DoSimdDot<u32x4, s16x8>();
-    case O::I16X8DotI8X16I7X16S: return DoSimdDot<u16x8, s8x16>();
+    case O::I16X8DotI8X16I7X16S: return DoSimdDot<s16x8, s8x16>();
```

Comment:
Without this change one of the relaxed simd tests fails under UBSAN:

```
- test/spec/relaxed-simd/relaxed_dot_product.txt                      
  expected error code 0, got 1.
  STDERR MISMATCH:
  --- expected
  +++ actual
  @@ -0,0 +1,2 @@
  +/usr/local/google/home/sbc/dev/wasm/wabt/src/interp/interp.cc:2472:30: runtime error: signed integer overflow: 65408 * 65409 cannot be represented in type 'int'
  +SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /usr/local/google/home/sbc/dev/wasm/wabt/src/interp/interp.cc:2472:30 in 

[+6|-1|%100] (0.13s) test/spec/relaxed-simd/relaxed_min_max.txt     
**** FAILED ******************************************************************
- test/spec/relaxed-simd/relaxed_dot_product.txt
    /usr/local/google/home/sbc/dev/wasm/wabt/bin/spectest-interp out/test/spec/relaxed-simd/relaxed_dot_product/relaxed_dot_product.json --enable-relaxed-simd

```

I'm honestly not sure why only this `Dot` opcode needs to change.  If I change the other 2 to also be signed then some tests in the simd repo start failing.  @ngzhian does this make any sense to you?

Reply:
C++ does multiplication in `int`, see https://en.cppreference.com/w/cpp/language/implicit_conversion#Integral_promotion

`u16*u16` behaves as if `i32*i32`

Reply:
It looks to me like `DoSimdDot` is indeed doing explicit casts on the multiplies: https://github.com/WebAssembly/wabt/blob/0df6c26d2a6122afcc2c32ff6b608326258d40d9/src/interp/interp.cc#L2472-L2473

That seems like its being very specific already.. so I'm a little confused.

Reply:
I guess the problem is that the inputs are s8 and the output is u16 (without this change) and when you do `u16(-128) * u16(-127)` you are multiplying two very large u16s.

Reply:
right, there is also the issue of sign extension, indeed.

Reply:
It looks like casting the result of the multiple rather than LHS / RHS does the trick.   Now the types remain unchanged and all tests pass!